src/Avalonia.Animation/Keyframes/KeyFramesStateMachine`1.cs

```diff
@@ -0,0 +1,276 @@
+using System;
+using System.Linq;
+using Avalonia.Data;
+
+namespace Avalonia.Animation.Keyframes
+{
+    /// <summary>
+    /// Provides statefulness for an iteration of a keyframe group animation.
+    /// </summary>
+    internal class KeyFramesStateMachine<T> : IObservable<object>, IDisposable
+    {
+        object _lastInterpValue;
+        object _firstKFValue;
+        
+        private ulong _delayTotalFrameCount,
+            _durationTotalFrameCount,
+            _delayFrameCount,
+            _durationFrameCount,
+            _repeatCount,
+            _currentIteration;
+
+        private bool _isLooping,
+                     _isRepeating,
+                     _isReversed,
+                     _checkLoopAndRepeat,
+                     _gotFirstKFValue;
+        private FillMode _fillMode;
+        private PlaybackDirection _animationDirection;
+        private KeyFramesStates _currentState, _savedState;
+        private KeyFrames<T> _parent;
+        private Animation _targetAnimation;
+        private Animatable _targetControl;
+        internal bool _unsubscribe = false;
+        private IObserver<object> _targetObserver;
+
+        private enum KeyFramesStates
+        {
+            INITIALIZE,
```

Comment:
I know these are private, but `ALL_CAPS` enums are sooo java ;)

Reply:
Hahaha right sir, will fix this one. I was in the influence of C++ when i was writing this one :laughing: